File: client_examples/python/serial_map.py
import io
import json

KEY_TYPE = b'\x10'
VALUE_TYPE = b'\x11'


class SerialMap:
    data: dict

    def __init__(self, data: bytes = None):
        self.data = {}

        if data is not None:
            buff = io.BytesIO(data)

            while buff.readable():
                if buff.read(1) != KEY_TYPE:
                    return
                len = int.from_bytes(buff.read(1), "big", signed=False)
                key = buff.read(len).decode("utf8")

                if buff.read(1) != VALUE_TYPE:
                    return
                len = int.from_bytes(buff.read(1), "big", signed=False)
                value = buff.read(len).decode("utf8")

                self.data[key] = value

    def put(self, key: str, value: str):
        self.data[key] = value

    def has(self, key: str) -> bool:
        return key in self.data

    def get(self, key: str) -> str:
        return self.data[key]

    def serialize(self) -> bytes:
        result = io.BytesIO()

        for key, value in self.data.items():
            result.write(KEY_TYPE)
            result.write(len(key).to_bytes(1, "big", signed=False))
            result.write(bytes(key, "utf8"))

            result.write(VALUE_TYPE)
            result.write(len(value).to_bytes(1, "big", signed=False))
            result.write(bytes(value, "utf8"))

        return result.getvalue()


def test():
    map = SerialMap()

    map.put("the key", "the value")
    print("Dictionary:", json.dumps(map.data))
    print("Serialized:", map.serialize())

    map2 = SerialMap(
        b'\x10\x05hello\x11\x05world\x10\x07another\x11\x05value')
    print("Serialized:", b'\x10\x05hello\x11\x05world\x10\x07another\x11\x05value')
    print("Dictionary:", json.dumps(map2.data))

File: client_examples/javascript/serialmap.js
class SerialMap {
    KEY_TYPE = 0x10;
    VALUE_TYPE = 0x11;
    /**
     * @type {[key:string]: string}
     */
    data = {};

    /**
     *
     * @param {Buffer} data
     */
    constructor(data = null) {
        if (data != null) {
            let cursor = 0;

            while (cursor < data.length) {
                if (data.readUInt8(cursor++) != this.KEY_TYPE) {
                    return;
                }
                let sz = data.readUInt8(cursor++);
                const key = data.slice(cursor, cursor + sz).toString("utf-8");
                cursor += sz;
                if (data.readUInt8(cursor++) != this.VALUE_TYPE) {
                    return;
                }
                sz = data.readUInt8(cursor++);
                const value = data.slice(cursor, cursor + sz).toString("utf-8");
                cursor += sz;

                this.data[key] = value;
            }
        }
    }

    /**
     *
     * @param {string} key
     * @param {string} value
     */
    put(key, value) {
        this.data[key] = value;
    }

    /**
     *
     * @param {string} key
     * @returns {string} The value
     */
    get(key) {
        return this.data[key];
    }

    /**
     *
     * @param {string} key
     * @returns {boolean} If the key exists
     */
    has(key) {
        return key in this.data;
    }

    /**
     * @returns {Buffer} The serialized buffer
     */
    serialize() {
        let size = 1;
        for (const [key, value] of Object.entries(this.data)) {
            size += key.length + value.length + 4;
        }

        let cursor = 0;
        const buf = Buffer.alloc(size);
        for (const [key, value] of Object.entries(this.data)) {
            cursor = buf.writeUInt8(this.KEY_TYPE, cursor);
            cursor = buf.writeUInt8(key.length, cursor);
            cursor += buf.write(key, cursor, "utf8");
            cursor = buf.writeUInt8(this.VALUE_TYPE, cursor);
            cursor = buf.writeUInt8(value.length, cursor);
            cursor += buf.write(value, cursor, "utf8");
        }

        cursor = buf.writeUInt8(0, cursor);

        return buf;
    }
}

module.exports = SerialMap;
